### PR TITLE
Option for arg/return type hints and correct typing for std::filesystem::path

### PR DESCRIPTION
**Pull request by timohl**
## Description

This PR adds the option for defining `arg_name` and `return_name` in a custom `type_caster`, which allows having different python type hints for arguments and return value. To check if `arg_name` or `return_name` is available or `name` should be used as fallback is implemented using the template classes `as_arg_type` and `as_return_type`.

Being entirely optional, this should not impact any existing `type_caster`.
This type hint specialization system is applied to the `type_caster` for `std::filesystem::path`.
Here, `Union[os.PathLike, str, bytes]` is the argument type hint and `Path` is the return type hint.
This is more accurate than the previous usage of `os.PathLike` for both argument and return type hint.

Also, most classes in pybind11::typing now support this feature for nested types, e.g., `py::typing::List<std::filesystem::path>` becomes `List[Union[os.PathLike, str, bytes]]` in arguments and `List[Path]` in return types.

I have added unit tests to check for the new signatures of stl/filesystem in typing containers and also to verify that `name` is used as fallback for other types (tested for `std::vector<std::filesystem::path>`).

## Suggested changelog entry:

```rst
Added option for different arg/return type hints to type_caster.
Updated stl/filesystem to use correct arg/return type hints.
Updated pybind11::typing to use correct arg/return type hints for nested types.
```

## Possible TODOs
* [x] Update `Custom type casters` documentation
* [x] ~~Update type hints for Eigen (currently uses `np.ndarray` as arg type hint but also takes lists and tuples (should probably be `numpy.typing.ArrayLike` + maybe `typing.Annotated` for shape/dtype annotation~~ (going into follow-up PR)

## Tests for py::typing classes:
* [x] Test for `std::vector<T>` (fallback)
* [x] Test for `Tuple<T...>`
* [x] Test for `Tuple<T, ellipsis>`
* [x] Test for `Dict<K, V>`
* [x] Test for `List<T>`
* [x] Test for `Set<T>`
* [x] Test for `Iterable<T>`
* [x] Test for `Iterator<T>`
* [x] Test for `Callable<Return(Args...)>`
* [x] Test for `Callable<Return(ellipsis)>`
* [x] Test for `Union<T...>`
* [x] Test for `Optional<T>`
* [x] Test for `TypeGuard<T>`
* [x] Test for `TypeIs<T>`

I would love to get feedback on this!
Especially on the "Possible TODO" regarding Eigen. Should I implement that here in this PR, or should I open a separate PR later?

Currently, I am working on adding typing stubs to Open3D (isl-org/Open3D#6917) using `pybind11-stubgen`.
Applying mypy/pyright on existing example code showed that a lot of type checks failed, which could be fixed by this PR.

### CODE REVIEW

File: include/pybind11/cast.h [hunk 1]
```diff
@@ -34,6 +34,39 @@ PYBIND11_WARNING_DISABLE_MSVC(4127)
...
+struct is_descr : std::false_type {};
+
+template <size_t N, typename... Ts>
+struct is_descr<descr<N, Ts...>> : std::true_type {};
+
+template <size_t N, typename... Ts>
+struct is_descr<const descr<N, Ts...>> : std::true_type {};
+
+// Use arg_name instead of name when available
+template <typename T, typename Enable = void>
```
Comment (rwgk) [inline include/pybind11/cast.h:48]: For consistency with existing pybind11 code, could you please make this `typename SFINAE = void`?

We already have 38 uses of `SFINAE`, but no `Enable`.

File: include/pybind11/typing.h [hunk 2]
```diff
@@ -131,6 +131,13 @@ struct handle_type_name<typing::Tuple<Types...>> {
     static constexpr auto name = const_name("tuple[")
                                  + ::pybind11::detail::concat(make_caster<Types>::name...)
                                  + const_name("]");
+    static constexpr auto arg_name
```
Comment (rwgk) [inline include/pybind11/typing.h:134]: How many of the added `arg_name` and `return_name` are covered by unit tests?

**Review by rwgk — COMMENTED** [review]
Nice!
 
> Update Custom type casters documentation

Could you please try to include that in this PR? — For awareness, please see [this comment](https://github.com/pybind/pybind11/pull/5416#discussion_r1813962686). Not sure if that changes anything for you. If not, completely fine.

> Update type hints for Eigen

Best in a follow-on PR.

File: include/pybind11/typing.h (hunk 2 above)
Comment (timohl) [reply include/pybind11/typing.h:134]: Currently just for 3 classes (+ one test for list of lists) but I can add tests for all.
I added an overview of implemented tests vs changes classes in my top comment (under Possible TODOs).
Should I keep adding all those tests to the filesystem test function or split them off into their own test function?

File: include/pybind11/cast.h (hunk 1 above)
Comment (timohl) [reply include/pybind11/cast.h]: Done

File: include/pybind11/typing.h (hunk 2 above)
Comment (rwgk) [reply include/pybind11/typing.h:134]: Hm ... up to you.

Splitting off sounds best, if that's not too much trouble.

**Comment by timohl** [comment]
> > Update Custom type casters documentation
> 
> Could you please try to include that in this PR? — For awareness, please see [this comment](https://github.com/pybind/pybind11/pull/5416#discussion_r1813962686). Not sure if that changes anything for you. If not, completely fine.

Sure! I can also try to adopt some of the documentation updates from #3862 as you suggested in #5416 since I am already at it.

**Comment by timohl** [comment]
While thinking about a test function for `TypeGuard` and `TypeIs`, I realized, that it probably should behave differently.
So, I changed it in [9ad445d](https://github.com/pybind/pybind11/pull/5450/commits/9ad445d93b55d1d3863fa10132a4f7e617811ef8) to always use the return type of the nested type when available.
Both are used only in return type hints and narrow down the type of the first argument.
Since return type hints are usually narrower than argument types, I think return type is more appropriate here.
(If any typing expert disagrees, please let me know)

Outstanding task: "Update Custom type casters documentation"
I will probably do that tomorrow.

**Comment by rwgk** [comment]
> One CI check failed, but that is a dependency timeout (maybe you can rerun it?).

Done and it passed (very quickly).

I can look at the code changes only later (might be a day or two).

File: tests/test_docs_advanced_cast_custom.cpp [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        // Decrement references (the tuple now owns x an y)
+        Py_DECREF(x);
+        Py_DECREF(y);
+        return t;
+    }
+
+    // Python -> C++: convert a `PyObject` into a `Point2D` and return false upon failure. The
+    // second argument indicates whether implicit conversions should be allowed.
+    bool load(handle src, bool) {
+        // Check if handle is valid Sequence of length 2
```
Comment (rwgk) [inline tests/test_docs_advanced_cast_custom.cpp:56]: Similarly here ...

Could you please try to use the pybind11 APIs?

Compare with `array_caster::convert_elements()` in pybind11/stl.h:

Roughly (totally untested):

```
        auto l = py::reinterpret_borrow<sequence>(seq);                             
        if (l.size() != 2) {                                                 
            return false;                                                       
        }
        double xy[2];
        ix = 0;                                                                      
        for (auto it : l) {                                                     
            py::make_caster<double> conv;                                                    
            if (!conv.load(it, convert)) {                                      
                return false;                                                   
            }
            xy[ix++] = py::cast_op<double>(conv);                                                                  
        }                                                                       
        value.x = xy[0];
        value.y = xy[1];
        return true;                                                            
```

Comment (timohl) [reply tests/test_docs_advanced_cast_custom.cpp]: I have changed it to use the pybind11 API.

File: tests/test_docs_advanced_cast_custom.cpp [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            Py_XDECREF(x);
+            Py_XDECREF(y);
+            return nullptr;
+        }
+        // Create tuple from x and y
+        auto *t = PyTuple_Pack(2, x, y);
+        // Decrement references (the tuple now owns x an y)
+        Py_DECREF(x);
+        Py_DECREF(y);
+        return t;
```
[the remark this answers is not in the document]
Comment (timohl) [reply tests/test_docs_advanced_cast_custom.cpp]: I have changed it to use the pybind11 API.

**Comment by timohl** [comment]
I have changed the example to only use the pybind11 API.
Additionally, I have changed the test for `py::typing` classes to only use pybind11 API for the caster.

The documentation text was also changed a little bit:
* Added recommendation to use pybind11 API to warning about Python C API
* Added note about `return_value_policy` and `convert` arguments referring to their respective sections in the documentation

File: docs/advanced/cast/custom.rst [hunk 5]
```diff
@@ -38,56 +56,82 @@ type is explicitly allowed.
...
+
+    template <>
+    struct type_caster<user_space::Point2D> {
+        // This macro inserts a lot of boilerplate code and sets the default type hint to `tuple`
+        PYBIND11_TYPE_CASTER(user_space::Point2D, const_name("tuple"));
+        // `arg_name` and `return_name` may optionally be used to specify type hints separately for
+        // arguments and return values.
+        // The signature of our negate function would then look like:
+        // `negate(Sequence[float]) -> tuple[float, float]`
+        static constexpr auto arg_name = const_name("Sequence[float]");
```
Comment (rwgk) [inline docs/advanced/cast/custom.rst:70]: `Sequence[int|float]` or `Sequence[Union[int, float]]`?

File: docs/advanced/cast/custom.rst [hunk 6]
```diff
@@ -38,56 +56,82 @@ type is explicitly allowed.
...
+                return false;
+            }
+            auto seq = py::reinterpret_borrow<py::sequence>(src);
+            // Check if exactly two values are in the Sequence
+            if (seq.size() != 2) {
+                return false;
+            }
+            // Check if each element is either a float or an int
+            for (auto item : seq) {
+                if (!py::isinstance<py::float_>(item) and !py::isinstance<py::int_>(item)) {
```
Comment (rwgk) [inline docs/advanced/cast/custom.rst:97]: Oversight: `and` -> `&&`

**Review by rwgk — APPROVED** [review]
Awesome work, thanks a lot!

@cryos could you help with a second set of eyes? Is there anything that stands out to you when glancing through?

File: docs/advanced/cast/custom.rst (hunk 6 above)
Comment (timohl) [reply docs/advanced/cast/custom.rst]: Fixed

File: docs/advanced/cast/custom.rst (hunk 5 above)
Comment (timohl) [reply docs/advanced/cast/custom.rst:70]: According to [PEP 484 – Type Hints](https://peps.python.org/pep-0484/#the-numeric-tower) and the [typing documentation](https://typing.readthedocs.io/en/latest/spec/special-types.html#special-cases-for-float-and-complex):

> when an argument is annotated as having type float, an argument of type int is acceptable

So, if we want a Sequence of floats and/or ints, I think, `Sequence[float]` is correct and shorter.
But I do not mind if you prefer to include `int`, as it is still correct, but just adds a little redundancy.

Comment (timohl) [reply docs/advanced/cast/custom.rst:70]: Although, now I see that I also used `Union[float,int]` in the `RealNumber` caster for the tests in [tests/test_pytypes.cpp](https://github.com/pybind/pybind11/pull/5450/files#diff-20bf1cca35b241c87fbf4ced7a3b55fd2421389f6861d06f67415c32d8c6bbc6).
However, I think that is ok there to just test the behavior of `arg_name`/`return_name`.
If you can think of a more sensible test class, I am open.

Comment (rwgk) [reply docs/advanced/cast/custom.rst:70]: > According to PEP 484 – Type Hints and the typing documentation:
>> when an argument is annotated as having type float, an argument of type int is acceptable

I wasn't aware of that, thanks for the explanation.

> However, I think that is ok there to just test the behavior of `arg_name`/`return_name`.

Could you maybe add a terse comment, that `int` isn't really needed, but appears for testing purposes? — For people who may use the test as a starting point for their work.

But looks good to me as is.

Comment (timohl) [reply docs/advanced/cast/custom.rst:70]: I have added a comment

Comment (rwgk) [reply docs/advanced/cast/custom.rst:70]: Perfect, thanks!

**Comment by timohl** [comment]
Made a small correction in the `docs/advanced/cast/overview` section in [6ea4704](https://github.com/pybind/pybind11/pull/5450/commits/6ea4704d538a9744568d92be7dead2f384626c3c).
I have searched the docs and this should be the only mention of `os.PathLike`, `pathlib.Path` or `std::filesystem::path`.

**Comment by rwgk** [comment]
Since there was no feedback from anyone else, I looked through the entire PR again. I found only one tiny oversight (aa21ab5294a406e602e468c14e0b3dd53c0c1341).

I also merged master.

Waiting for CI to finish. When I see it passing I'll merge.

**Comment by HolyBlackCat** [comment]
`static constexpr auto return_name = const_name("Path");` seems to be wrong, `Path` isn't a valid type? I've reported https://github.com/pybind/pybind11/issues/5477

**Pull request merged**